[AMDGPU] HasOneUse uses

jofrn opened the pull request:
#91578 implements `HasOneUse` predicate on `PatFrag`, so this commit uses it within AMDGPU.

Pierre-vh commented on llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:
```diff
@@ -468,7 +468,23 @@ bool GIMatchTableExecutor::executeMatchTable(
+      assert(MI->getNumDefs() > 0 && "No defs");
+      const Register Res = MI->getOperand(0).getReg();
+
+      if (!MRI.hasOneNonDBGUse(Res))
+        if (handleReject() == RejectAndGiveUp)
+          return false;
```
```suggestion
      if (!MRI.hasOneNonDBGUse(Res)) {
        if (handleReject() == RejectAndGiveUp)
          return false;
      }
```

Pierre-vh commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -884,6 +884,9 @@ class PatFrags<dag ops, list<dag> frags, code pred = [{}],
   // If set to true, a predicate is added that checks for the absence of use of
   // the first result.
   bit HasNoUse = ?;
+  // If set to true, a predicate is added that checks for the sole use of
+  // the first result.
```
```suggestion
  // If set to true, a predicate is added that checks for that the first result only has one non-debug user.
```

jofrn replied:
dag emitter checks `hasOneUse()` though. This file is generic to both emitters.

Also, `GIM_CheckHasNoUse` is implemented with `use_nodbg_empty`, yet the comment above `bit HasNoUse = ?;` has non-debug users as implicit.

arsenm replied:
I wouldn't bother spelling out the debug user case; it shouldn't be relevant much longer since debug instructions are going away

jofrn replied:
Ok. Changed it back to something similar yet without mentioning that. thank you.

Pierre-vh commented on llvm/utils/TableGen/Common/GlobalISel/GlobalISelMatchTable.h:
```diff
@@ -1691,6 +1692,28 @@ class NoUsePredicateMatcher : public InstructionPredicateMatcher {
   }
 };
 
+/// Generates code to check for the sole use of the result.
```
```suggestion
/// Generates code to check that the first result has only one non-debug user.
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUInstructions.td:
```diff
@@ -169,29 +169,22 @@ def brtarget   : Operand<OtherVT>;
-  let GISelPredicateCode = [{
-    return MRI.hasOneNonDBGUse(MI.getOperand(0).getReg());
-  }];
+  [{ return true; }]> {
```
Can't we just remove the argument entirely?

arsenm replied:
Yes, the custom predicates should go

jofrn committed:
[TableGen] HasOneUse builtin predicate on PatFrags

This predicate tells GlobalISelEmitter and DAGISelEmitter to check that the instruction to emit has only one use of its result. This can be used on a PatFrag instead of defining custom predicates for both emitters per record that requires it.

jofrn committed:
[AMDGPU] HasOneUse uses

jofrn commented on llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:
```diff
@@ -468,7 +468,23 @@ bool GIMatchTableExecutor::executeMatchTable(
+
+      if (!MRI.hasOneNonDBGUse(Res))
+        if (handleReject() == RejectAndGiveUp)
+          return false;
```
```
if (!MRI.use_nodbg_empty(Res)) {
  if (handleReject() == RejectAndGiveUp)
    return false;
}
```
is how HasNoUse is implemented. Shouldn't we keep these styles the same?

jofrn committed:
Merge branch 'main' into hasoneuse-uses
